axelor-account/src/main/java/com/axelor/apps/account/service/invoice/InvoiceServiceImpl.java

```diff
@@ -974,4 +980,129 @@ protected long getNotLetteredAdvancePaymentMoveLinesAmount(Partner partner) {
+        AccountConfig accountConfig = accountConfigRepository.findByCompany(invoice.getCompany());
+        if (invoice.getFinancialDiscount().getDiscountBaseSelect()
+                == FinancialDiscountRepository.DISCOUNT_BASE_VAT
+            && (invoice.getOperationTypeSelect()
+                    == InvoiceRepository.OPERATION_TYPE_SUPPLIER_PURCHASE
+                || invoice.getOperationTypeSelect() == InvoiceRepository.OPERATION_TYPE_CLIENT_SALE)
+            && accountConfig.getPurchFinancialDiscountTax() != null) {
+          return invoice
+              .getInTaxTotal()
+              .multiply(invoice.getFinancialDiscountRate().divide(new BigDecimal(100), 3))
+              .multiply(accountConfig.getPurchFinancialDiscountTax().getActiveTaxLine().getValue())
+              .divide(
+                  accountConfig
+                      .getPurchFinancialDiscountTax()
+                      .getActiveTaxLine()
+                      .getValue()
+                      .add(new BigDecimal(1)),
+                  5,
+                  RoundingMode.CEILING);
+        } else if (invoice.getFinancialDiscount().getDiscountBaseSelect()
+                == FinancialDiscountRepository.DISCOUNT_BASE_VAT
+            && (invoice.getOperationTypeSelect() == InvoiceRepository.OPERATION_TYPE_SUPPLIER_REFUND
+                || invoice.getOperationTypeSelect()
+                    == InvoiceRepository.OPERATION_TYPE_CLIENT_REFUND)
+            && accountConfig.getSaleFinancialDiscountTax() != null) {
+          return invoice
+              .getInTaxTotal()
+              .multiply(invoice.getFinancialDiscountRate().divide(new BigDecimal(100), 3))
+              .multiply(accountConfig.getSaleFinancialDiscountTax().getActiveTaxLine().getValue())
+              .divide(
+                  accountConfig
+                      .getSaleFinancialDiscountTax()
+                      .getActiveTaxLine()
+                      .getValue()
+                      .add(new BigDecimal(1)),
+                  5,
+                  RoundingMode.CEILING);
+        }
+      }
+    }
+    return BigDecimal.ZERO;
+  }
+
+  public BigDecimal calculateFinancialDiscountTotalAmount(Invoice invoice) {
+    return calculateFinancialDiscountAmount(invoice)
+        .add(calculateFinancialDiscountTaxAmount(invoice));
+  }
+
+  public BigDecimal calculateAmountRemainingInPayment(Invoice invoice, Boolean apply) {
+    if (apply && invoice != null) {
+      return invoice
+          .getAmountRemaining()
+          .subtract(calculateFinancialDiscountTaxAmount(invoice))
+          .subtract(calculateFinancialDiscountAmount(invoice));
+    }
+    return invoice.getAmountRemaining();
+  }
+
+  public boolean applyFinancialDiscount(Invoice invoice) {
+    if (invoice != null) {
+      if (invoice.getFinancialDiscountDeadlineDate() != null
+          && invoice.getFinancialDiscountDeadlineDate().compareTo(LocalDate.now()) >= 0) {
```

Comment:
to replace by a call to appBaseService.getTodayDate() to manage timezone and make tests easier.